File: ATOM.py
from math import sqrt
from ruckig import InputParameter, Result, Ruckig, Trajectory  # pip install ruckig


class State():
	def __init__(self, v_idx, ang_idx, disc_vels, irr_times, is_final=False):
		self.v_idx = v_idx
		self.ang_idx = ang_idx
		self.v = disc_vels[v_idx]
		self.vel_square = self.v * self.v
		if is_final:
			assert self.v == 0
			self.half_window = 0
		else:
			self.half_window = self.v * irr_times[self.ang_idx] * 0.5

		self.key = str(self.v_idx) + "_" + str(ang_idx)

	def __repr__(self):
		return self.key

	def __str__(self):
		return self.key

	def __hash__(self):
		return hash(self.key)


def linspace(a, b, n):
	assert a < b
	assert n > 1
	delta = (b - a) / (n-1)

	return [a + i*delta for i in range(n)]


def calc_time_between_segments(irr_times, switch_times, angle_distances, max_window, idx, v0, v1, parameters):
	assert idx > 0
	idx0 = idx - 1
	idx1 = idx

	window_size_0 = v0 * irr_times[idx0]
	window_size_1 = v1 * irr_times[idx1]

	if window_size_0 > max_window or window_size_1 > max_window:
		return float('inf')
	remaining_angle = angle_distances[idx0] - (window_size_0+window_size_1)/2
	assert remaining_angle > 0

	elst = switch_times[idx0]

	otg = Ruckig(1)
	inp = InputParameter(1)
	trajectory = Trajectory(1)

	inp.current_position = [0.0]
	inp.current_velocity = [v0]
	inp.current_acceleration = [0.0]
	
	inp.target_velocity = [v1]
	inp.target_position = [remaining_angle]
	inp.target_acceleration = [0.0]
	
	inp.max_velocity = [parameters["v_max"]]
	inp.max_acceleration = [parameters["a_max"]]
	inp.max_jerk = [parameters["j_max"]]

	inp.min_velocity = [-1.0e-16]  # This would be 0, but there are some rounding errors that cause issues, it seems.
	inp.min_acceleration = [parameters["a_min"]]

	inp.minimum_duration = elst
	if not otg.validate_input(inp, check_current_state_within_limits=True, check_target_state_within_limits=True):
		return float('inf')

	# Calculate the trajectory in an offline manner
	try:
		result = otg.calculate(inp, trajectory)
	except RuntimeError:
		return float('inf')
	if result == Result.ErrorInvalidInput:
		return float('inf')

	return trajectory.duration + irr_times[idx0]


def get_neighs(all_states, current, has_been_visited):
	# TODO: This is a disgrace. Make it faster.
	return list(filter(lambda el: not has_been_visited[current.ang_idx+1][all_states[current.ang_idx+1].index(el)], all_states[current.ang_idx+1]))


def find_current(open_set, f_score):
	# TODO: This might be too slow. Use a more clever data structure. A fibonacci heap, probably. 
	best_state = None
	best_f_score = float('inf')
	for state in open_set:
		if f_score[state.ang_idx][state.v_idx] < best_f_score:
			best_f_score = f_score[state.ang_idx][state.v_idx]
			best_state = state
		elif f_score[state.ang_idx][state.v_idx] == best_f_score:
			if state.v > best_state.v:
				best_f_score = f_score[state.ang_idx][state.v_idx]
				best_state = state

	assert best_f_score < float('inf')
	return best_state


def atom(irr_times, elsts, angle_distances, maximum_window_size, parameters, vel_res=256):
	n = len(irr_times)
	assert n > 1
	assert n - 1 == len(elsts)
	assert n - 1 == len(angle_distances)
	assert maximum_window_size < min(angle_distances)
	assert min(irr_times) >= 0
	assert min(elsts) >= 0
	assert vel_res > 1

	disc_vels = linspace(0, parameters["v_max"], vel_res)
	max_v = parameters["v_max"]
	max_acc = parameters["a_max"]
	min_acc = parameters["a_min"]

	assert max_v > 0
	assert min_acc < 0
	assert max_acc > 0
	assert parameters["j_max"] > 0

	heuristic_from_ang_idx = [sum(irr_times[i:]) - irr_times[-1] + sum(elsts[i:]) for i in range(n)]

	came_from = {}

	g_score = []
	f_score = []

	all_states = []
	initial_state = State(0, 0, disc_vels, irr_times)
	final_state = State(0, n - 1, disc_vels, irr_times, is_final=True)
	all_states.append([initial_state])
	has_been_visited = []
	g_score.append([0])
	f_score.append([heuristic_from_ang_idx[initial_state.ang_idx]])
	has_been_visited.append([False])
	for ang_idx in range(1, n - 1):
		states_for_angle = []
		has_been_visited_for_angle = []
		g_score_for_angle = []
		f_score_for_angle = []
		for v_idx in range(len(disc_vels)):
			state = State(v_idx, ang_idx, disc_vels, irr_times)
			states_for_angle.append(state)
			has_been_visited_for_angle.append(False)
			g_score_for_angle.append(float('inf'))
			f_score_for_angle.append(float('inf'))
		g_score.append(g_score_for_angle)
		f_score.append(f_score_for_angle)
		has_been_visited.append(has_been_visited_for_angle)
		all_states.append(states_for_angle)

	has_been_visited.append([False])
	all_states.append([final_state])
	g_score.append([float('inf')])
	f_score.append([float('inf')])

	open_set = set([initial_state])

	found_it = False
	while len(open_set) > 0:
		current = find_current(open_set, f_score)
		assert not has_been_visited[current.ang_idx][current.v_idx]
		has_been_visited[current.ang_idx][current.v_idx] = True
		if current == final_state:
			assert heuristic_from_ang_idx[current.ang_idx] == 0
			found_it = True
			break

		open_set.remove(current)
		current_g_score = g_score[current.ang_idx][current.v_idx]

		neighs = get_neighs(all_states, current, has_been_visited)
		for neigh in neighs:
			assert neigh.ang_idx == current.ang_idx + 1
		if len(neighs) > 0:
			angle_dist_between_center_points = angle_distances[current.ang_idx]

			local_max_vel = sqrt(2 * max_acc * (angle_dist_between_center_points - current.half_window) + current.vel_square)
			local_min_vel = 0 if 2 * min_acc * (angle_dist_between_center_points - current.half_window) + current.vel_square < 0 else sqrt(2 * min_acc * (angle_dist_between_center_points - current.half_window) + current.vel_square)

			g_scores_neighs = g_score[neighs[0].ang_idx]
			f_scores_neighs = f_score[neighs[0].ang_idx]
			elst = elsts[current.ang_idx]
			irr_time_from = irr_times[current.ang_idx]
			for neigh_state in neighs:
				if neigh_state.v > local_max_vel or neigh_state.v < local_min_vel:
					continue
				d = calc_time_between_segments(irr_times, elsts, angle_distances, maximum_window_size, neigh_state.ang_idx, current.v, neigh_state.v, parameters)
				if d != float('inf'):
					tentative_g_score = current_g_score + d
					if tentative_g_score <= g_scores_neighs[neigh_state.v_idx]:
						came_from[neigh_state] = current
						g_scores_neighs[neigh_state.v_idx] = tentative_g_score
						f_scores_neighs[neigh_state.v_idx] = tentative_g_score + heuristic_from_ang_idx[neigh_state.ang_idx]

						if has_been_visited[neigh_state.ang_idx][neigh_state.v_idx] == False and neigh_state not in open_set:
							open_set.add(neigh_state)

	assert found_it
	time_val = g_score[current.ang_idx][current.v_idx] + irr_times[-1]
	traj = [current.v_idx]
	while current in came_from:
		traj.append(max_v * came_from[current].v_idx / (vel_res - 1))
		current = came_from[current]
	traj = list(reversed(traj))
	return time_val, traj

File: ATOM.cpp
#include <iostream>
#include <unordered_map>
#include <unordered_set>
#include <ruckig/ruckig.hpp>

#include <chrono>
#include <cmath>
#include <utility>
#include <vector>
#include <cassert>
#include <set>
#include <limits>
#include <random>

using namespace std::chrono;
using namespace ruckig;

namespace {
#define VEL_DISC 256

class State {
  public:
	size_t vIdx;
	size_t angIdx;
	double v;
	double velSquare;

	size_t key;
	double halfWindow;

	State() :
		vIdx(0u),
		angIdx(0u),
		v(0.0),
		velSquare(0.0),
		key(0u),
		halfWindow(0.0)
	{};

	State(size_t vIdx_, size_t angIdx_, double v_, double irrTime) :
		vIdx(vIdx_),
		angIdx(angIdx_),
		v(v_),
		velSquare(v_ * v_),
		key(vIdx + angIdx * VEL_DISC),
		halfWindow(v_ * irrTime * 0.5 )
	{}

	bool operator==(const State& other) const {
		return key == other.key;
	}

	size_t operator() () const {
		return key;
	}

	bool operator< (const State& right) const {
		return key < right.key;
	}
};

double calcCost(
    const std::vector<double>& irrTimes,
    double elst,
    const std::vector<double>& angleDistances,
    double maximumWindowSize,
    size_t neighAngIdx,
    double currentV,
    double neighV,
    double vMax,
    double aMax,
    double jMax) {
	assert(neighAngIdx >= 1);
	const size_t currentAngleIdx = neighAngIdx - 1;

	const auto windowSize0 = currentV * irrTimes.at(currentAngleIdx);
	const auto windowSize1 = neighV * irrTimes.at(neighAngIdx);

	if (windowSize0 > maximumWindowSize || windowSize1 > maximumWindowSize) {
		return std::numeric_limits<double>::max();
	}

	const double remainingAngle = angleDistances[currentAngleIdx] -  0.5 * (windowSize0 + windowSize1);
	assert(remainingAngle > 0);

	InputParameter<1> input;
	input.current_position = { 0.0 };
	input.current_velocity = { currentV };
	input.current_acceleration = { 0.0 };

	input.target_position = { remainingAngle };
	input.target_velocity = { neighV };
	input.target_acceleration = { 0.0 };

	input.max_velocity = { vMax };
	input.max_acceleration = { aMax };
	input.max_jerk = { jMax };

	input.min_velocity = { -vMax};
	input.minimum_duration = elst;

	// We don't need to pass the control rate (cycle time) when using only offline features
	Ruckig<1> otg;
	Trajectory<1> trajectory;

	// Calculate the trajectory in an offline manner (outside of the control loop)
	const Result result = otg.calculate(input, trajectory);
	if (result != Result::Working ) {
		return std::numeric_limits<double>::max();
	}
	assert(result != Result::Finished);
	assert(trajectory.get_duration() >= elst - 0.00001);
	return trajectory.get_duration() + irrTimes.at(currentAngleIdx);
}

State findCurrent(std::unordered_map<size_t, State>& stateSet, std::set<std::pair<double, size_t>>& stateSetF) {
	size_t keyOfBest = stateSetF.begin()->second;
	const State bestStateCopy = stateSet.find(keyOfBest)->second;

	stateSet.erase(keyOfBest);
	stateSetF.erase(stateSetF.begin());
	return bestStateCopy;
}

std::vector<double> getDiscVels(double vMax) {
	std::vector<double> discVels(VEL_DISC);
	for ( size_t i = 0; i < VEL_DISC; ++i) {
		discVels.at(i) = i * vMax / (VEL_DISC - 1);
	}
	return discVels;
}
}

double atom(
    size_t n,
    std::vector<double>& irrTimes,
    std::vector<double>& elsts,
    std::vector<double>& angleDistances,
    double maximumWindowSize,
    double vMax,
    double aMax,
    double jMax) {
	const auto discVels = getDiscVels(vMax);

	double cumsumIrr = 0.0;
	double cumsumElst = 0.0;

	assert(!irrTimes.empty());
	assert(n > 0);
	const auto irrSum = std::accumulate( irrTimes.cbegin(), --irrTimes.cend(), 0.0 );  // NOTE: We're skipping the last element.
	const auto elstSum = std::accumulate(elsts.begin(), elsts.end(), 0.0);

	std::vector<double> heuristicFromAngIdx(n);
	for (size_t i = 0; i < n; ++i) {
		heuristicFromAngIdx.at(i) = elstSum + irrSum - cumsumIrr - cumsumElst;
		if (i + 1 != n) {
			cumsumElst += elsts.at(i);
			cumsumIrr += irrTimes.at(i);
		}
	}

	auto initialState = State(0, 0,  0.0, irrTimes.front());
	auto finalState = State(0, n - 1,  0.0, irrTimes.at(n - 1));

	std::vector<std::vector<double>> gScores = { {0.0} };
	std::vector<std::vector<double>> fScores = { {heuristicFromAngIdx.at(initialState.angIdx)} };
	std::vector<std::vector<State>> allStates = { {initialState} };
	std::vector<std::vector<bool>> hasBeenVisited = { {false} };

	for (size_t angIdx = 1; angIdx < n - 1; angIdx++) {
		std::vector<double> gScoreForAngle(VEL_DISC, std::numeric_limits<double>::max());
		std::vector<double> fScoreForAngle(VEL_DISC, std::numeric_limits<double>::max());
		std::vector<State> statesForAngle;
		statesForAngle.reserve(VEL_DISC);
		std::vector<bool> hasBeenVisitedForAngle(VEL_DISC, false);
		for (size_t vIdx = 0; vIdx < VEL_DISC; vIdx++) {
			auto state = State(vIdx, angIdx, discVels.at(vIdx), irrTimes.at(angIdx));
			statesForAngle.push_back(state);
		}
		gScores.push_back(gScoreForAngle);
		fScores.push_back(fScoreForAngle);
		hasBeenVisited.push_back(hasBeenVisitedForAngle);
		allStates.push_back(statesForAngle);
	}

	hasBeenVisited.push_back({ false });
	allStates.push_back({ finalState });

	gScores.push_back({ std::numeric_limits<double>::max() });
	fScores.push_back({ std::numeric_limits<double>::max() });

	std::unordered_map<size_t, State> stateSet = {{0, initialState}};
	std::set<std::pair<double, size_t>> stateSetF = {{ fScores.at(0).at(0), 0 }};

	std::unordered_map<size_t, State> bestPrev;
	std::set<std::pair<double, size_t>> openSetSorted = { {double{0}, size_t{0}} };

	bool foundIt = false;
	while ( !stateSet.empty()) {
		const auto currentState = findCurrent(stateSet, stateSetF);
		hasBeenVisited.at(currentState.angIdx).at(currentState.vIdx) = true;

		if (currentState.key == finalState.key) {
			foundIt = true;
			break;
		}
		assert(currentState.angIdx < n);
		assert(currentState.angIdx < gScores.size());
		assert(currentState.vIdx < gScores.at(currentState.angIdx).size());
		auto currentGScore = gScores.at(currentState.angIdx).at(currentState.vIdx);

		const auto angleDistBetweenCenterPoints = angleDistances.at(currentState.angIdx);
		const auto localMinVel =  -2.0 * aMax * (angleDistBetweenCenterPoints - currentState.halfWindow) + currentState.velSquare < 0.0 ?
		                          0.0  :
		                          sqrt( -2.0  * aMax * (angleDistBetweenCenterPoints - currentState.halfWindow) + currentState.velSquare);
		const auto localMaxVel = sqrt(2.0 * aMax * (angleDistBetweenCenterPoints - currentState.halfWindow) + currentState.velSquare);

		const auto elst = elsts.at(currentState.angIdx);
		const size_t a = currentState.angIdx + 1;
		auto& gScoresNeighs = gScores.at(a);
		auto& fScoresNeighs = fScores.at(a);

		for ( size_t i = 0; i < allStates.at(a).size(); ++i) {
			if ( hasBeenVisited.at(a).at(i)) {
				continue;
			}
			const auto& neighState = allStates.at(a).at(i);
			if (neighState.v < localMinVel) {
				continue;
			}
			if ( neighState.v > localMaxVel) {
				break;
			}

			const auto dist = calcCost(irrTimes,
			                           elst,
			                           angleDistances,
			                           maximumWindowSize,
			                           neighState.angIdx,
			                           currentState.v,
			                           neighState.v,
			                           vMax,
			                           aMax,
			                           jMax);

			if (dist < std::numeric_limits<double>::max()) {
				const auto tentativeGScore = currentGScore + dist;
				assert(neighState.vIdx < gScoresNeighs.size());

				if (tentativeGScore <= gScoresNeighs.at(neighState.vIdx) ) {
					assert(neighState.angIdx < heuristicFromAngIdx.size());

					const auto oldF = fScoresNeighs.at(neighState.vIdx);

					gScoresNeighs.at(neighState.vIdx) = tentativeGScore;
					fScoresNeighs.at(neighState.vIdx) = tentativeGScore + heuristicFromAngIdx.at(neighState.angIdx);

					assert(neighState.angIdx < hasBeenVisited.size());
					assert(neighState.vIdx < hasBeenVisited.at(neighState.angIdx).size());
					assert(!hasBeenVisited.at(neighState.angIdx).at(neighState.vIdx));
					if (stateSet.count(neighState.key) == 0) {
						stateSet.insert({ neighState.key, neighState });
					} else {
						stateSetF.erase({ oldF, neighState.key });
					}
					stateSetF.emplace(fScoresNeighs.at(neighState.vIdx), neighState.key);
					bestPrev.insert({ allStates.at(a).at(i).key, currentState });
				}
			}
		}
	}
	auto currentStateKey = finalState.key;
	while (bestPrev.count(currentStateKey) == 1) {
		const auto& currentState = bestPrev.at(currentStateKey);
		currentStateKey = currentState.key;
	}

	assert(foundIt);
	auto timeVal = gScores.at(finalState.angIdx).at(finalState.vIdx) + irrTimes.back();
	return timeVal;
}

int main() {
	srand(1);

	milliseconds ms = duration_cast<milliseconds>(
	                      system_clock::now().time_since_epoch());
	std::default_random_engine generator(0ULL);

	auto s = 0.0;
	for (size_t iter = 0; iter < 10; iter++) {
		size_t n_ = 360;
		auto ELST_down_ =  0.5;
		auto ELST_up_ =  10 * ELST_down_;
		std::vector<double> angleDistances_(n_ - 1, 2.0);
		auto maxWindow_ =  2.0;

		std::uniform_double_distribution<double> distributionElst(0.0, 1.0);
		std::uniform_double_distribution<double> distributionIrrs(0.0, 1.26);

		std::vector<double> irrTimes_(n_);
		for (size_t i = 0; i < n_; i++) {
			irrTimes_.at(i) = distributionIrrs(generator);
		}

		std::vector<double> switchTimes_(n_ - 1);
		for (size_t i = 0; i < n_ - 1; i++) {
			if (distributionElst(generator) <  0.1 ) {
				switchTimes_.at(i) = ELST_up_;
			} else {
				switchTimes_.at(i) = ELST_down_;
			}
		}

		auto t2 = atom(n_, irrTimes_, switchTimes_, angleDistances_, maxWindow_,  5.0, 0.5, 0.5);
		std::cout << t2 << "," <<
		          std::accumulate(irrTimes_.begin(), irrTimes_.end(), 0.0) + std::accumulate(switchTimes_.begin(), switchTimes_.end(), 0.0)
		          << std::endl;
		for (auto el : irrTimes_) {
			std::cout << el << ", ";
		}
		std::cout << "\n";
		for (auto el : switchTimes_) {
			std::cout << el << ", ";
		}
		std::cout << "\n";
		s += t2;
	}

	milliseconds ms2 = duration_cast<milliseconds>(
	                       system_clock::now().time_since_epoch());

	std::cout << "Time taken to run a for loop = " << ms2.count() - ms.count() << " milliseconds.\n";
}
